dali/pipeline/executor/executor2/exec_graph.h

```diff
@@ -0,0 +1,385 @@
+
+  /** Limits the number of output buffers for the operator.
+   *
+   * If the graph is scheduled multiple times ahead, the operator would produce multiple results,
+   * potentially consuming a lot of memory. The output_queue_limit limits the number of pending
+   * outputs. Once all consumers are done with previous iterations, the semaphore is released
+   * and the task may proceed with the next iteration.
+   */
+  std::shared_ptr<tasking::Semaphore> output_queue_limit;
+
+  /** The instance of the operator (or null for output node) */
+  const std::unique_ptr<OperatorBase> op;
+
+  /** Data-independent execution environment (thread pool, stream, etc). */
+  ExecEnv env = {};
+
+  /** Obtains a worskpace from a workspace cache or, if not found, creates a new one. */
+  std::pair<std::unique_ptr<Workspace>, SharedEventLease> GetWorkspace(WorkspaceParams params);
+
+  void PutWorkspace(std::unique_ptr<Workspace> &&ws);
+
+  /** The instance name of the operator. */
+  const std::string instance_name;
+
+  /** The backend on which the operator runs. */
+  const OpType backend = OpType::CPU;
+
+  /** Whether the node is the very output of the pipeline. There's only one such node. */
+  const bool is_pipeline_output = false;
+
+  /** Whether the operator in the node is a batch size provider. */
+  const bool is_batch_size_provider = false;
+
+  /** Visit marker for graph algorithms. */
+  mutable bool visited = false;
+
+ private:
+  /** The task from the previous iteration - kept in order to maintain execution order */
+  tasking::SharedTask prev_task_;
+
+  /** The task from the current iteration */
+  tasking::SharedTask main_task_;
+
+  /** The task that releases the output_queue_limit semaphore.
+   *
+   * IMPORTANT: Release_outputs_ is NOT required for correctness. It's sole function is limiting
+   *            the number of pending outputs from the operator.
+   *
+   * This is an auxiliary task which is scheduled after all direct successors of main_task_.
+   *
+   * op A --- (output 0) ----- op B --------- ...
+   *    \                         /
+   *     ------- (output 1) -----(- op C --- ...
+   *                              \   |
+   *                               \  |
+   *                                \  \
+   *                                 \_release outptus (A)
+   */
+  tasking::SharedTask release_outputs_;
+
+  /** Creates a workspace at the pipeline's output.
+   *
+   * Output node's inputs are converted to Pipeline's outputs. The output workspace doesn't have
+   * any inputs.
+   */
+  std::unique_ptr<Workspace> CreateOutputWorkspace();
+  /** Creates a workspace suitable for running the operator `op`. */
+  std::unique_ptr<Workspace> CreateOpWorkspace();
+
+  /** The workspace.
+   *
+   * An operator never executes in parallel with its previous instance,
+   * so we never need more than one workspace.
+   */
+  std::unique_ptr<Workspace> ws_;
+  bool has_workspace_ = false;
```

Comment:
why not simply checking if `ws_` is null?

Reply:
This flag indicates that the workspace is _somewhere_. We get it and put it back. It indicates that the workspace has been created. It's for bug checks, basically.

Reply:
I extended the function description to include the intended usage of GetWorkspace/PutWorkspace.